docs/Outputs.md

```diff
@@ -0,0 +1,225 @@
+For example there will be `json.encode`, as well as `csv.encode` for encoders, and `http.post` and `tcp.send` for transports.
+
+The minimal set that must be implemented is `http.post` and `json.encode`.
+
+Endpoints are implemented in pure Flux.
+
+## Flux
+
+Encoders are provided in Flux as `encode` functions.
+For example the JSON encoder is part of the `json` package.
+Encoders take an object `data` and return its encoding as a string:
+
+```flux
+encode: (data:object) -> string
+```
+
+Communication protocols are also implemented (in Go), so that one can, for example, send an HTTP request with JSON data in pure Flux:
+
+```flux
+import "http"
+import "json"
+
+data = {a: 1, b: 2}
+json_enc = json.encode(data)
+http.post(header: {"Content Type": "application-json"}, url: "http://some/url/", data: json_enc)
+```
+
+External services in Flux provide an `endpoint` function for building an endpoint.
+The `endpoint` function returns a factory function for building a specialized transformation for sending tables to the external service.
+The factory function is used to create multiple ways of sending data to the service, as such it accepts a `mapFn` to map each record in a table to a proper object.
+The fields in that object are used by the endpoint to build the message for the external service.
+Every external service must define how that object must be shaped for the communication to happen.
+
+For example:
+
+```flux
+import ext "external_service"
+
+someFilter = (r) => ...
+someOtherFilter = (r) => ...
+
+// Obtain the factory function with this configuration for the endpoint.
+endpoint = ext.endpoint(url: "https://url/to/external/service")
+
+someMapFn = (r) => {
+   // Generate payload from record.
+   return {title: r.someTag, body: string(v: r._value)}
+})
+someOtherMapFn = (r) => {
+   // Generate payload from record.
+   return {title: r.someOtherTag, body: string(v: r._value)}
+})
+
+// Invoke the factory to get custom transformations.
+to1 = endpoint(mapFn: someMapFn)
+to2 = endpoint(mapFn: someOtherMapFn)
+
+from(...)
+    |> range(...)
+    |> filter(fn: someFilter)
+    |> to1()
+
+from(...)
+    |> range(...)
+    |> filter(fn: someOtherFilter)
+    |> to2()
+```
+
+The output of `to` functions contains the result of the communication with the external service and varies on a per-service base.
+
+
+### An Example: Slack
+
+We provide below an example implementation for a Slack endpoint.
+The package provides an `endpoint` function that return the Slack endpoint.
+It also provides helpers to send a single message in a channel and build the request header.
+
+```flux
+package slack
+
+import "http"
+import "json"
+
+// This can be globally overridden in another package with:
+// slack.defaultURL = "https://slack.com/api/chat.anotherURL"
+option defaultURL = "https://slack.com/api/chat.postMessage"
+
+// Util function to create the header.
+header = (token, ctype="application/json") => ({
```

Comment:
The functions in the package make up the public API of the package, since `header` is not needed publicly and is only called in one location, I would prefer to see this logic moved into the `message` function instead of exposing it directly.